scripts/rv_expression_association/prepare_inputs.py

```diff
@@ -0,0 +1,211 @@
+)  # 'v0/plink_files/vpreb3_rare_promoter.bim'
+@click.option(
+    '--genotype-file-fam', required=True
+)  # 'v0/plink_files/vpreb3_rare_promoter.fam'
+@click.option(
+    '--phenotype-file', required=True
+)  # 'scrna-seq/grch38_association_files/expression_files/B_naive_expression.tsv'
+@click.option('--kinship-file', required=True)  # 'v0/skat/grm_wide.csv'
+def prepare_inputs(  # pylint: disable=missing-function-docstring, too-many-locals
+    cell_type: str,
+    gene_name: str,
+    sample_mapping_file: str,
+    genotype_file_bed: str,
+    genotype_file_bim: str,
+    genotype_file_fam: str,
+    phenotype_file: str,
+    kinship_file: str,
+):
+
+    expression_filename = AnyPath(output_path(f'{gene_name}_{cell_type}.csv'))
+    genotype_filename = AnyPath(output_path(f'{gene_name}_rare_regulatory.csv'))
+    kinship_filename = AnyPath(output_path('kinship_common_samples.csv'))
+
+    # region PHENOTYPE_FILE
+
+    phenotype = pd.read_csv(phenotype_file, sep='\t', index_col=0)
+
+    phenotype = xr.DataArray(
+        phenotype.values,
+        dims=['sample', 'gene'],
+        coords={'sample': phenotype.index.values, 'gene': phenotype.columns.values},
+    )
+
+    # endregion PHENOTYPE_FILE
+
+    # region KINSHIP_FILE
+
+    # read in GRM (genotype relationship matrix; kinship matrix)
+    kinship = pd.read_csv(kinship_file, index_col=0)
+    kinship.index = kinship.index.astype('str')
+    assert all(kinship.columns == kinship.index)  # symmetric matrix, donors x donors
+
+    kinship = xr.DataArray(
+        kinship.values,
+        dims=['sample_0', 'sample_1'],
+        coords={'sample_0': kinship.columns, 'sample_1': kinship.index},
+    )
+    kinship = kinship.sortby('sample_0').sortby('sample_1')
+
+    # endregion KINSHIP_FILE
+
+    # region GENOTYPE_FILE
+
+    # read in genotype file (plink format)
+    # bed
+    with to_path(genotype_file_bed).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.bed', 'wb') as handle:
+        handle.writelines(data)
+    # bim
+    with to_path(genotype_file_bim).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.bim', 'wb') as handle:
+        handle.writelines(data)
+    # fam
+    with to_path(genotype_file_fam).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.fam', 'wb') as handle:
+        handle.writelines(data)
+    # read
+    geno = read_plink1_bin('temp.bed')
+
+    # endregion GENOTYPE_FILE
+
+    # region SAMPLE_MAPPING_FILE
+
+    # this file will map different IDs (and OneK1K ID to CPG ID)
+    sample_mapping = pd.read_csv(sample_mapping_file, sep='\t')
+
+    # samples with expression data
+    donors_onek1k = sample_mapping['OneK1K_ID'].unique()
+    donors_onek1k.sort()
+    donors_exprs = sorted(
+        set(list(phenotype.sample.values)).intersection(donors_onek1k)
+    )
+    logging.info(f'Number of unique donors with expression data: {len(donors_exprs)}')
```

Comment:
```suggestion
    # samples with expression data
    donors_exprs = set(phenotype.sample.values).intersection(
        set(sample_mapping['OneK1K_ID'].unique())
    )
    
    logging.info(f'Number of unique donors with expression data: {len(donors_exprs)}')
```

I think there are some extra steps here that aren't helping. `unique()` returns a series, which can be cast directly as a list/set.

This is being used downstream in a set intersection, so the ordering doesn't matter, in which case set operations are faster, and you can skip a few steps.

Reply:
i can't remember fully, but i think there was a reason i was sorting them for later? I agree the first `.sort()` is useless, but would sorting the final set make sense? i.e. `sorted(set(...))`

Reply:
`sorted` turns the set into a list. Later this is used in the `is_in()` operation, where a set would be faster, and the ordering won't alter the results.

e.g. as far as I'm aware when you do

```
sample_mapping.loc[sample_mapping['OneK1K_ID'].isin(donors_exprs)]
```

the ordering of the results is based on `sample_mapping`, not `donors_exprs`

Reply:
ok! Ideally I'd like to return all objects (y, geno, and kinship) not only with the same samples, but in the same order. This is complicated (for my limited abilities) by the fact that the samples are called differently in each of the objects, so arguably I should just change them all to the same? Now i deal with this very uglily (and hopefully correct???) in R, but a better solution would be more than welcome

Reply:
Aye, definitely makes sense to maintain the ordering where possible. The issue here would be that the objects you're sorting don't ultimately alter the order of the DataFrame. 

I try and stay away from R where possible :D 

It should be simple in Pandas to make sure the ID column is always titled `CPG_ID` then sort the DF on those values to make sure it's consistent across objects

Reply:
sorry I misspoke - it is not that the column name is different, but rather that the IDs themselves are different - in the expression file, the OneK1K_IDs are used (e.g., 943_944), in the genotype the CPG ones are (e.g., CPG9951)